Tenant Token (#89)
meilisearch/specifications: text/0089-tenant-tokens.md

- The only rule at the moment is the search parameter `filter`. Other rules may be added in the future.
- When a request is made with the Tenant Token, MeiliSearch checks if this Tenant Token is authorized to make the request and in this case injects the rules at search time.

### 1.2 Motivation

`Tenant tokens` are introduced to solve multi-tenant use-cases.

Users today need to set up workarounds to have multi-tenant indexes. Most of the time, they have to use server code to implement the access restriction logic before requesting MeiliSearch. It is difficult to maintain, to implement, and the performance is degraded because the frontend code does not communicate directly with MeiliSearch.

### 1.3 `Tenant Token` Explanations

#### 1.3.1 Example: Solving Multi-Tenancy with `Tenant tokens`

![](https://user-images.githubusercontent.com/3692335/149373631-de6f3c5f-a514-4c8d-b018-ee09ccaeaf4d.png)

`Mark` is a developer for a SaaS platform. He would like to ensure that every end-user can only access their documents at search time.

> When an end-user registers, the Mark's backend code generates a `Tenant token` for that end-user so they can only access their documents.

Reviewer: ```suggestion
When an end-user registers, Mark's backend code generates a `Tenant token` for that end-user so they can only access their documents.
```

This tenant-token is signed with a MeiliSearch API Key so that MeiliSearch can ensure that the tenant-token has been generated from a known entity.

> MeiliSearch check if the Tenant Token is authorized to make the search request.
> 
> Then MeiliSearch extract the Tenant Token's rules to apply for the search request.

Reviewer: ```suggestion
MeiliSearch checks if the Tenant Token is authorized to make the search request.

Then MeiliSearch extracts the Tenant Token's rules to apply for the search request.
```

##### 2.1.2.3 Payload example

Given a MeiliSearch API Key used to sign the JWT from the user code. Here is an example of a valid payload for a tenant token.

e.g `MeiliSearch API key: rkDxFUHd02193e120218f72cc51a9db62729fdb4003e271f960d1631b54f3426fa8b2595`

```json
{
    "apiKeyPrefix": "rkDxFUHd", <- The first 8 characters of the signing MeiliSearch API Key
    "exp": 1641835850, <- An expiration date in seconds from 1970-01-01T00:00:00Z UTC
    "indexesPolicy": { <- The indexesPolicy Json Object.
        "*": {
            "filter": "user_id = 1"
        }
    }
}
```

> In this example, `indexesPolicy` allows to specify, that no matter which index is searched (among all those accessible by the signing API key that generated the tenant token), this filter will be applied on all search requests.

##### 2.1.2.4 `apiKeyPrefix` field

> `apiKeyPrefix` permits to verify that the signing API key of the Token is known and valid within MeiliSearch. It must contain the first 8 characters of the MeiliSearch API key that generates and signs the Tenant Token.

Reviewer 2: ```suggestion
`apiKeyPrefix` contains the 8 first characters of the API key that generated the tenant token. `apiKeyPrefix` is used to find the whole `apiKey` inside the stored API keys in MeiliSearch. Which in turn, validates the signature of the token.
```

##### 2.1.2.5 `exp` field

> `exp` permits to specify the expiration date of the Tenant Token if needed. The format is a JSON numeric value representing the number of seconds from 1970-01-01T00:00:00Z UTC until the specified UTC date/time, ignoring leap seconds.

Reviewer 2: ```suggestion
`exp` is used to specify the expiration date of the Tenant Token if needed. The format is a JSON numeric value representing the number of seconds from 1970-01-01T00:00:00Z UTC until the specified UTC date/time, ignoring leap seconds.
```

##### 2.1.2.6 `indexesPolicy` JSON object

> `indexesPolicy` is a description of the possible rules for each index.

Reviewer 2: ```suggestion
`indexesPolicy` contains the rules of search on different indexes.
```
`Description` implies more an explanation than a format.

Here are some valid examples in an attempt to cover all possible use cases.

---

> > In this case, all searchable indexes from the signing API Key will be searchable by the tenant token without specific rules.

Reviewer 2: ```suggestion
> In this case, all indexes on which the signing API Key has permissions are searchable by the tenant token without any restrictions.
```

```json
{
    "indexesPolicy": {
        "*": {}
    }
}
```

is equivalent to

```json
{
    "indexesPolicy": {
        "*": null
    }
}
```

is equivalent to

```json
{
    "indexesPolicy": ["*"] //This notation does not allow the addition of specific rules. The search will just be accessible on all accessibles indexes from the signing API Key for the Tenant Token without specific rules.
}
```

---

> > In this case, all searchable indexes from the signing API Key will be searchable by the tenant token and MeiliSearch will apply the filter definition before applying the search parameters added by the end user.

Reviewer 2: ```suggestion
> In this case, all indexes on which the signing API Key has permissions are searchable by the tenant token but the additional `filters` limit the search results of the token user. The filters are always applied and overwrite similar filters the end user might add to the search parameters. 
```

```json
{
    "indexesPolicy": {
        "*": {
            "filter": "user_id = 1"
        }
    }
}
```

---

> > In this case, if the medical_records index is searchable from the signing API Key, the tenant token can only search in the medical_records index without applying specific rules before applying the search parameters added by the end user.

Reviewer 2: ```suggestion
> In this case, if the `medical_records` index is searchable from the signing API Key, the tenant token can only search in the `medical_records` index. No further rules impact their search results on `medical_records`.
```

```json
{
    "indexesPolicy": {
        "medical_records": {}
    }
}
```
is equivalent to

```json
{
    "indexesPolicy": {
        "medical_records": null
    }
}
```

is equivalent to

```json
{
    "indexesPolicy": ["medical_records"]
}
```

---

> > In this case, if the medical_records index is searchable from the signing API Key, the tenant token can only search in the medical_records index and specific rules will be applied at search time before applying the search parameters added by the end user.

Reviewer 2: ```suggestion
> In this case, if the medical_records index is searchable from the signing API Key, the tenant token can only search in the medical_records index, plus specific rules limits their search results. These rules are always applied and cannot be overwritten by the users search parameters.
```

```json
{
    "indexesPolicy": {
        "medical_records": {
            "filter": "user_id = 1"
        }
    }
}
```

---

> > In this case, if the medical_records and medical_appointments indexes are searchable from the signing API Key, the tenant token can only search in those indexes and will apply specific rules given the searched index before applying the search parameters added by the end user.

Reviewer 2: ```suggestion
> In this case, if the medical_records and medical_appointments indexes are searchable from the signing API Key, the tenant token can only search in those indexes and specific rules limits their search results. These rules are always applied and cannot be overwritten by the users search parameters.
```

```json
{
    "indexesPolicy": {
        "medical_records": {
            "filter": "user_id = 1"
        },
        "medical_appointments": {
            "filter": "user_id = 1 AND accepted = true"
        }
    }
}
```

---

> > In this case, all searchable indexes from the signing API Key will be searchable and the rules under `*` will be applied at search time. The medical_appointments index policy will replace the `*` rules specifically when this index is searched by those specified for that index.

Reviewer 2: ```suggestion
> In this case, all searchable indexes from the signing API Key are searchable and have their search apply specific rules defined in the wildcard field: `*`. The `medical_appointments` index rules, defined in the field of the same name, overwrites the rules defined in the wildcard field`*`.
```

A Tenant Token is a JWT containing the information necessary for MeiliSearch to verify it and extract permission/rules to apply it to the end user's search.

#### 1.1.1 Summary Key Points

- `Tenant tokens` are JWTs generated on the user side. Thus not stored or retrievable on MeiliSearch side.
- `Tenant tokens` contain rules that ensure that a `Tenant token` holder (e.g an end-user) only has access documents matching rules chosen at the `tenant token` creation.
- `Tenant tokens` are signed from a MeiliSearch `API key` on the user's code.
- `Tenant tokens` cannot be less restrictive than the signing `API key` and can only be used for searching. A Tenant Token cannot search within more indexes than the API Key that signed that Tenant Token.
- `Tenant tokens` can have different rules for each index accessible by the signing API key. These filters rule per index are described in an `searchRules` json object.
- The only rule at the moment is the search parameter `filter`. Other rules may be added in the future.

> - When a request is made with the Tenant Token, MeiliSearch checks if this Tenant Token is authorized to make the request and in this case injects the rules at search time.

Reviewer 3: Can we make this more explicit @gmourier? Because reading that now, makes me think in two possibilities:

- the SDK will decode the JWT and then get the `searchRules` to send them into the search request
- the Meilisearch server will receive the tenant token and it should be responsible to decode + apply these search rules on the fly.

Author: I've added some precisions. Let me know if it's enough. 😇

        "*": {
            "filter": "user_id = 1"
        }
    }
}
```

---

> In this case, if the `medical_records` index is searchable from the signing API Key, the tenant token can only search in the `medical_records` index. No further rules impact search results on `medical_records`.

```json
{
    "searchRules": {
        "medical_records": {}
    }
}
```
is equivalent to

```json
{
    "searchRules": {
        "medical_records": null
    }
}
```

is equivalent to

```json
{
    "searchRules": ["medical_records"]
}
```

---

> In this case, if the `medical_records` index is searchable from the signing API Key, the tenant token can only search in the `medical_records` index, MeiliSearch will apply the filter rule before applying the search parameters added by the end user.

```json
{
    "searchRules": {
        "medical_records": {
            "filter": "user_id = 1"
        }
    }
}
```

---

> In this case, if the `medical_records` and `medical_appointments` indexes are searchable from the signing API Key, the tenant token can only search in those indexes. MeiliSearch will apply the filter rule before applying the search parameters added by the end user.

```json
{

>     "searchRules": {

Reviewer 4: Also, it could be good idea to validate a token, by performing a placeholder search with it.

>     "searchRules": {

Author: Some users have asked for the possibility to hide the `tenant` field that can be used in a `searchRules, which will probably be implemented in the future. With this type of error, an end-user could have the information of the field that allowed the multi-tenancy when it could have been explicitly hidden via a `hideAttributes` attribute.

The tenant tokens will be generated programmatically in the vast majority of cases. It might be useful to include the TenantToken source in the error message to make it clear that the error comes from `searchRules` format and thus facilitate error resolution, but I wonder how this would work for an end-user who is reporting the bug on a field that needs to be `hidden`.

>     "searchRules": {

Reviewer 3: Hey @gmourier I understood your suggestion, and I think it is not the best since we will be creating a connection between two different API's and like you said, blocking the future possibilities of this, but I understood that maybe not possible to add the source to the error message either.

I just wanted to raise the concern with you. What do you think to add some sentence in the `Future possibilities` regarding this topic? 
Talking to @curquiza she told me that is best to add a new [discussion in the product](https://github.com/meilisearch/product/discussions/398) in order to keep track of this.

>     "searchRules": {

Author: Thanks Bruno, I will add it to the Future Possibilities!
